add some known limitations about TiFlash (#5891)
pingcap/docs-cn: tiflash/use-tiflash.md

> * TiFlash 目前的 Decimal 除法计算上和 TiDB 有一点不兼容，具体体现在在进行 Decimal 相除的时候，TiFlash 会始终按照编译时推断出来的类型进行计算，而 TiDB 则在计算过程中采用高于编译是推断出来的类型，这导致在一些带有 Decimal 除法的 SQL 在 TiDB + TiKV 上的结果会和 TiDB + TiFlash 上的结果不一样。举例来说，对于如下的 case：
>
> ```
> mysql> create table t (a decimal(3,0), b decimal(10, 0));
> Query OK, 0 rows affected (0.07 sec)
> 
> mysql> insert into t values (43, 1044774912);
> Query OK, 1 row affected (0.03 sec)
> 
> mysql> alter table t set tiflash replica 1;
> Query OK, 0 rows affected (0.07 sec)
> 
> mysql> set session tidb_isolation_read_engines='tikv';
> Query OK, 0 rows affected (0.00 sec)
> 
> mysql> select a/b, a/b + 0.0000000000001 from t where a/b;
> +--------+-----------------------+
> | a/b    | a/b + 0.0000000000001 |
> +--------+-----------------------+
> | 0.0000 |       0.0000000410001 |
> +--------+-----------------------+
> 1 row in set (0.00 sec)
> 
> mysql> set session tidb_isolation_read_engines='tiflash';
> Query OK, 0 rows affected (0.00 sec)
> 
> mysql> select a/b, a/b + 0.0000000000001 from t where a/b;
> Empty set (0.01 sec)
> 
> ```
>

> > 在 TiDB 和 TiFlash 中，`a/b`在编译期推导出来的 Type 都为 `Decimal(7,4)`，而在`Decimal(7,4)`的约束小，`a/b`返回的结果应该为`0.0000`，但是在 TiDB 中，`a/b` 运行期的精度比 `Decimal(7,4)`大，所以原表中的数据没有被 `where a/b` 过滤掉，而在 TiFlash 中`a/b`在运行期也是采用`Decimal(7,4)`作为结果类型，所以原表中的数据被`where a/b`过滤掉了。

Reviewer: 注意事项内容太多，建议不用 note 格式，单独放一个标题写注意事项

Author: 好的。

在执行计划中，出现了 `ExchangeReceiver` 和 `ExchangeSender` 算子。该执行计划表示 `nation` 表读取完毕后，经过 `ExchangeSender` 算子广播到各个节点中，与 `customer` 表先后进行 `HashJoin` 和 `HashAgg` 操作，再将结果返回至 TiDB 中。

TiFlash 提供了两个全局/会话变量决定是否选择 Broadcast Hash Join，分别为：

- [`tidb_broadcast_join_threshold_size`](/system-variables.md#tidb_broadcast_join_threshold_count-从-v50-版本开始引入 )，单位为 bytes。如果表大小（字节数）小于该值，则选择 Broadcast Hash Join 算法。否则选择 Shuffled Hash Join 算法。
- [`tidb_broadcast_join_threshold_count`](/system-variables.md#tidb_broadcast_join_threshold_count-从-v50-版本开始引入 )，单位为行数。如果 join 的对象为子查询，优化器无法估计子查询结果集大小，在这种情况下通过结果集行数判断。如果子查询的行数估计值小于该变量，则选择 Broadcast Hash Join 算法。否则选择 Shuffled Hash Join 算法。

> ## 注意事项

Reviewer: 这里更像是使用限制？

Author: 感觉还是注意事项比较好，这里都是些corner case不支持，或者corner case和TiDB表现不一致的东西

> 在 TiFlash 目前有一些已知的不支持的功能以及与原生 TiDB 不兼容的问题，具体罗列如下：

Reviewer: ```suggestion
TiFlash 目前尚不支持的一些功能，与原生 TiDB 可能存在不兼容的问题，具体如下：
```

* TiFlash 计算层：

>     * TiDB 4.0.2 版本之前，不支持 TiDB 新排序规则框架，所以在 TiDB 开启[新框架下的排序规则支持](/character-set-and-collation.md#新框架下的排序规则支持)后不支持任何表达式的下推，TiDB 4.0.2 以及后续的版本取消了这个限制。

Reviewer: 建议在 dev 和 v5.0 版本删除这条信息，因为对 dev 和 v5.0 版本的用户没有实际意义。可以在 v4.0 版本文档中保留。
```suggestion
```

>     * 不支持从数值溢出检测，例如对于 2 个 `bigint` 最大值相加 `9223372036854775807 + 9223372036854775807`，在 TiDB 中应有的行为是提示错误 `ERROR 1690 (22003): BIGINT value is out of range`，但如果该计算在 TiFlash 中进行，则会得到溢出的结果 `-2`无报错。

Reviewer: ```suggestion
    * 不支持检查溢出的数值。例如将两个 `BIGINT` 类型的最大值相加 `9223372036854775807 + 9223372036854775807`，该计算在 TiDB 中预期的行为是返回错误 `ERROR 1690 (22003): BIGINT value is out of range`，但如果该计算在 TiFlash 中进行，则会得到溢出的结果 `-2` 且无报错。
```

>     * 不支持 Window Function。

Reviewer: ```suggestion
    * 不支持窗口函数。
```

    * 不支持从 TiKV 读取数据。

>     * TiFlash 目前的 sum 函数不支持参数为 String 类型的输入，但是因为一些原因，TiDB 在编译时无法检测出这种情况，所以当出现类似 `select sum(string_col) from t`的时候，TiFlash 会报`[FLASH:Coprocessor:Unimplemented] CastStringAsReal is not supported.`的错，要避免这类报错，需要手动把 SQL 改写成`select sum(cast(string_col as double)) from t`

Reviewer: ```suggestion
    * 目前 TiFlash 中的 `sum` 函数不支持传入字符串类型的参数，但 TiDB 在编译时无法检测出这种情况。所以当执行类似于 `select sum(string_col) from t` 的语句时，TiFlash 会报错 `[FLASH:Coprocessor:Unimplemented] CastStringAsReal is not supported.`。要避免这类报错，需要手动把 SQL 改写成 `select sum(cast(string_col as double)) from t`。
```

>     * TiFlash 目前的 Decimal 除法计算上和 TiDB 有一点不兼容，具体体现在在进行 Decimal 相除的时候，TiFlash 会始终按照编译时推断出来的类型进行计算，而 TiDB 则在计算过程中采用高于编译是推断出来的类型，这导致在一些带有 Decimal 除法的 SQL 在 TiDB + TiKV 上的结果会和 TiDB + TiFlash 上的结果不一样。举例来说，对于如下的 case：

Reviewer: ```suggestion
    * TiFlash 目前的 Decimal 除法计算和 TiDB 存在不兼容的情况。例如在进行 Decimal 相除的时候，TiFlash 会始终按照编译时推断出来的类型进行计算，而 TiDB 则在计算过程中采用精度高于编译时推断出来的类型。这导致在一些带有 Decimal 除法的 SQL 语句在 TiDB + TiKV 上的执行结果会和 TiDB + TiFlash 上的执行结果不一样，示例如下：
```

>     ```

Reviewer: ```suggestion
    ```sql
```

    mysql> create table t (a decimal(3,0), b decimal(10, 0));
    Query OK, 0 rows affected (0.07 sec)

    mysql> insert into t values (43, 1044774912);
    Query OK, 1 row affected (0.03 sec)

    mysql> alter table t set tiflash replica 1;
    Query OK, 0 rows affected (0.07 sec)

    mysql> set session tidb_isolation_read_engines='tikv';
    Query OK, 0 rows affected (0.00 sec)

    mysql> select a/b, a/b + 0.0000000000001 from t where a/b;
    +--------+-----------------------+
    | a/b    | a/b + 0.0000000000001 |
    +--------+-----------------------+
    | 0.0000 |       0.0000000410001 |
    +--------+-----------------------+
    1 row in set (0.00 sec)

    mysql> set session tidb_isolation_read_engines='tiflash';
    Query OK, 0 rows affected (0.00 sec)

    mysql> select a/b, a/b + 0.0000000000001 from t where a/b;
    Empty set (0.01 sec)

>     ```

Reviewer: 建议再缩进四个空格

Author: 已更新

>     在 TiDB 和 TiFlash 中，`a/b`在编译期推导出来的 Type 都为 `Decimal(7,4)`，而在`Decimal(7,4)`的约束下，`a/b`返回的结果应该为`0.0000`，但是在 TiDB 中，`a/b` 运行期的精度比 `Decimal(7,4)`大，所以原表中的数据没有被 `where a/b` 过滤掉，而在 TiFlash 中`a/b`在运行期也是采用`Decimal(7,4)`作为结果类型，所以原表中的数据被`where a/b`过滤掉了。

Reviewer: ```suggestion
    以上示例中，在 TiDB 和 TiFlash 中，`a/b` 在编译期推导出来的类型都为 `Decimal(7,4)`，而在 `Decimal(7,4)` 的约束下，`a/b` 返回的结果应该为 `0.0000`。但是在 TiDB 中，`a/b` 运行期的精度比 `Decimal(7,4)` 高，所以原表中的数据没有被 `where a/b` 过滤掉。而在 TiFlash 中 `a/b` 在运行期也是采用 `Decimal(7,4)` 作为结果类型，所以原表中的数据被 `where a/b` 过滤掉了。
```

  * TiFlash 目前的 Decimal 除法计算上和 TiDB 有一点不兼容，具体体现在在进行 Decimal 相除的时候，TiFlash 会始终按照编译时推断出来的类型进行计算，而 TiDB 则在计算过程中采用高于编译是推断出来的类型，这导致在一些带有 Decimal 除法的 SQL 在 TiDB + TiKV 上的结果会和 TiDB + TiFlash 上的结果不一样。举例来说，对于如下的 case：

```
mysql> create table t (a decimal(3,0), b decimal(10, 0));
Query OK, 0 rows affected (0.07 sec)

mysql> insert into t values (43, 1044774912);
Query OK, 1 row affected (0.03 sec)

mysql> alter table t set tiflash replica 1;
Query OK, 0 rows affected (0.07 sec)

mysql> set session tidb_isolation_read_engines='tikv';
Query OK, 0 rows affected (0.00 sec)

mysql> select a/b, a/b + 0.0000000000001 from t where a/b;
+--------+-----------------------+
| a/b    | a/b + 0.0000000000001 |
+--------+-----------------------+
| 0.0000 |       0.0000000410001 |
+--------+-----------------------+
1 row in set (0.00 sec)

mysql> set session tidb_isolation_read_engines='tiflash';
Query OK, 0 rows affected (0.00 sec)

mysql> select a/b, a/b + 0.0000000000001 from t where a/b;
Empty set (0.01 sec)

```

> 在 TiDB 和 TiFlash 中，`a/b`在编译期推导出来的 Type 都为 `Decimal(7,4)`，而在`Decimal(7,4)`的约束小，`a/b`返回的结果应该为`0.0000`，但是在 TiDB 中，`a/b` 运行期的精度比 `Decimal(7,4)`大，所以原表中的数据没有被 `where a/b` 过滤掉，而在 TiFlash 中`a/b`在运行期也是采用`Decimal(7,4)`作为结果类型，所以原表中的数据被`where a/b`过滤掉了。

Reviewer 2: “约束小”？

Author: 是“约束下”，已更新。。
